packages/app/src/game/module/chat/sagas.ts

```diff
@@ -0,0 +1,32 @@
+import { take, delay, select, put } from "redux-saga/effects";
+import { quickChatCommands } from ".";
+import { PlayerActions, PlayerEvents } from "../../../../../gamemode/lib";
+
+
+export const handleQuickChat = function*() {
+	while (true) {
+		const action = yield take(PlayerEvents.playerReceiveQuickChatEvent.toString());
+		const { sendingPlayerId, receivingPlayerId, chatValue } = action.payload;
+		if (!chatValue) {
+			return;
+		}
+		yield put(quickChatCommands.setPlayerChat({ sendingPlayerId, receivingPlayerId, chatValue }));
+		let chatToCheck = chatValue;
+		while (true) {
```

Comment:
So I think here rather than using this `while` loop you could structure the saga a little differently, and then use a `race` effect to wait for the delay/ a new message. What a `race` effect does is allow you to start 2 or more effects (there is a `race` method for promises too btw) and then simply get the result from the one that finishes soonest. In this case we will either want to wait 3000ms and clear the chat, or wait until a new chat comes in and do it immediately

https://redux-saga.js.org/docs/advanced/RacingEffects/

We'll also need to differentiate the first `take`, because that one won't be part of a race effect. I would do it something like this (pseudocode):

```ts
// take the first quick chat event
let action = yield take(PlayerEvents.playerReceiveQuickChatEvent.toString())

while (true) {
    // use the current value of `action`
    const { sendingPlayerId, /* ... */ } = action.payload;
    yield put(setPlayerChat(/* ... */));

    // start the take() and the delay(), and then either `newChat` or `timeout` will be set depending on
    // which finishes first. the other will be cancelled, and the value will be `undefined`
    const { newChat, timeout } = yield race({
    	newChat: take(PlayerEvents.playerReceiveQuickChatEvent.toString()),
	timeout: delay(3000)
    })

    // if there was a new chat message received, then store it in `action` and start the
    // loop again to take care of it
    if (newChat) {
	action = newChat;
	continue;
    }

    // otherwise, the timeout happened before a new chat
    yield put(clearChat());
   
    // and now wait for a new chat to come back
    // this will then wait for the `take` to be fulfilled, set it to `action` and then immediately
    // start this loop again
    action = yield take(PlayerEvents.playerReceiveQuickChatEvent.toString())
}
```

Reply:
Damn, that's elegant ❤